### ISSUE TITLE
Bind dynamic named arguments

### ISSUE THREAD

**Issue by jameschenjav** [quietboil/sibyl#10]
Hi there,

I need to build a dynamic query based on users input. For example, there are 2 possible arguments `foo_from` and `foo_to`. I will generate the where clauses based on how many args provided. So the query will be one of above:

```sql
SELECT * FROM bar WHERE foo >= :foo_from AND bar <= :foo_to;  -- provided both
SELECT * FROM bar WHERE bar <= :foo_to;  -- only foo_to
SELECT * FROM bar WHERE foo >= :foo_from;  -- only foo_from
SELECT * FROM bar;  -- none
```

I tried to build the bind `args` with `Vec`, but I can't convert it to a tuple. I also tried to implement trait `ToSql`, but some of the params in `bind_to` are using symbols from private modules (like `stmt: &oci::OCIStmt`), which made me unable to implement it by myself.

Currently, I have to write 4 different queries with 4 different bindings, just because the empty string is the same as `NULL` in Oracle. Is there anyway to dynamically bind any number of named arguments? Or is there any plan to make it possible?

**Comment by quietboil** [comment]
Depending on the type of the arguments you might be able to work around this even now as sibyl knows how to bind slices of primitive numerics.

> *Note* that `u8` is a special case as it not treated as a sequence of values, but rather as raw data.

For example:

```rust
let stmt = session.prepare("
    SELECT location_id, state_province, city, postal_code, street_address
      FROM hr.locations
     WHERE location_id = :id1 OR location_id = :id2
     ORDER BY location_id
")?;

let rows = stmt.query( [2500, 2600].as_ref() )?;

while let Some(row) = rows.next()? {
    let location_id    : i32  = row.get(0)?;
    let state_province : &str = row.get(1)?;
    let city           : &str = row.get(2)?;
    let postal_code    : &str = row.get(3)?;
    let street_address : &str = row.get(4)?;
    match location_id {
        2500 => {
            assert_eq!(state_province, "Oxford");
            assert_eq!(city, "Oxford");
            assert_eq!(postal_code, "OX9 9ZB");
            assert_eq!(street_address, "Magdalen Centre, The Oxford Science Park");
        },
        2600 => {
            assert_eq!(state_province, "Manchester");
            assert_eq!(city, "Stretford");
            assert_eq!(postal_code, "09629850293");
            assert_eq!(street_address, "9702 Chester Road");
        },
        _ => {
            panic!("unexpected location");
        }
    }
}
let num_rows = stmt.row_count()?;
assert_eq!(num_rows, 2);
```

You can have other arguments in there too and even name them:

```rust
let stmt = session.prepare("
    SELECT location_id, state_province, city, postal_code, street_address
      FROM hr.locations
     WHERE country_id = :country
       AND (location_id = :id OR location_id = :id2)
       AND postal_code LIKE :code
     ORDER BY location_id
")?;
let rows = stmt.query( (
    ("COUNTRY", "UK"),
    ("ID", [2500, 2600].as_ref()),
    ("CODE", "%9%")
) )?;
```

Unfortunately, this is as far as it can go *today*. I'll need to add similar capabilities for other types.

**Comment by quietboil** [comment]
Just released 0.6.16 has previously extended support for passing slices as consecutive arguments to types that were not covered in 0.6.12 and before. 0.6.16 also added passing arguments that are collected into a vector. See `dyn_args.rs` and `dyn_args_slice.rs` examples.
